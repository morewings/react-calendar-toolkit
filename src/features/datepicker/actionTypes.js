export const SET_DATE = 'SET_DATE';
