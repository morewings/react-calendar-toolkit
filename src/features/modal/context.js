import {createContext, useContext} from 'react';

const ModalContext = createContext({});

export const useModalContext = () => useContext(ModalContext);

export default ModalContext;
