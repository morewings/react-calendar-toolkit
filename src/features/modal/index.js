export {default as ModalReducer, initialState} from './ModalReducer';
export {default as ModalContext, useModalContext} from './context';
export {default as useModalActions} from './actionCreators';
