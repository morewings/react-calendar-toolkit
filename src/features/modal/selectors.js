export const getShowDatepicker = state => state.modal.showDatepicker;
