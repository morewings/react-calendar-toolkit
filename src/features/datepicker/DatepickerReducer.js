import {SET_DATE, SET_TODAY, SET_PRECISION} from './actionTypes';

const initialState = {
  selectedTimestamp: 0,
  todayTimestamp: 0,
  precision: 'day',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_DATE: {
      const {selectedTimestamp, precision} = action.payload;
      return {
        ...state,
        selectedTimestamp,
        precision,
      };
    }
    case SET_TODAY: {
      const {todayTimestamp} = action.payload;
      return {
        ...state,
        todayTimestamp,
      };
    }
    case SET_PRECISION: {
      return {
        ...state,
        precision: action.payload,
      };
    }
    default:
      return state;
  }
};
