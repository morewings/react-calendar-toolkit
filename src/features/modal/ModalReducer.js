import {TOGGLE_DATEPICKER} from './actionTypes';

export const initialState = {
  isVisible: false,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case TOGGLE_DATEPICKER: {
      const isVisible = action.payload;
      return {
        ...state,
        isVisible,
      };
    }
    default:
      return state;
  }
};
