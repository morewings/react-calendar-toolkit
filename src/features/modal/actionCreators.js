import {useModalContext} from './context';
import {TOGGLE_DATEPICKER} from './actionTypes';

const toggleDatePicker = isVisible => ({
  type: TOGGLE_DATEPICKER,
  payload: isVisible,
});

const useModalActions = () => {
  const {dispatch} = useModalContext();
  return {
    toggleDatePicker: isVisible => {
      dispatch(toggleDatePicker(isVisible));
    },
  };
};

export default useModalActions;
