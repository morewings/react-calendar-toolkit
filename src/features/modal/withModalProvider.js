import React from 'react';
import createStoreProvider from 'store';
import ModalContext from './context';
import ModalReducer, {initialState} from './ModalReducer';

export const Provider = createStoreProvider({
  initialState,
  reducer: ModalReducer,
  context: ModalContext,
});

const withModalProvider = WrappedComponent => props => (
  <Provider>
    <WrappedComponent {...props} />
  </Provider>
);

export default withModalProvider;
