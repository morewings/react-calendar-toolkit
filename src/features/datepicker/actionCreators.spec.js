import React from 'react';
import {renderHook, act} from '@testing-library/react-hooks';
import createStoreProvider from 'store';
import useModalActions from 'features/modal/actionCreators';
import {TOGGLE_DATEPICKER} from 'features/modal/actionTypes';
import {matchDatesWithPrecision} from 'utils/dateUtils';
import {
  SET_DATE,
  SET_PRECISION,
  SET_TODAY,
  SET_VISIBILITY,
} from './actionTypes';
import useDatePickerActions from './actionCreators';
import {initialState} from './DatepickerReducer';
import DatePickerContext from './context';

const mockReducer = jest.fn();

const Provider = createStoreProvider({
  initialState,
  reducer: mockReducer,
  context: DatePickerContext,
});

const date = new Date('2020-01-20T00:00:00.000'); // 20.01.2020

const setDatePayload = {
  selectedTimestamp: 1579478400000,
};

const setTodayPayload = {
  todayTimestamp: 1579478400000,
};

const setVisibilityPayload = {
  visibleTimestamp: 1579478400000,
};

const precisionPayload = 'precision';

describe('useModalActions', () => {
  beforeEach(() => {
    mockReducer.mockClear();
  });

  it('returns action creators', () => {
    const {result} = renderHook(() => useDatePickerActions(), {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });
    expect(result.current).toMatchSnapshot();
  });

  describe.each`
    action             | type              | value               | expectedPayload
    ${'setDate'}       | ${SET_DATE}       | ${date}             | ${setDatePayload}
    ${'setToday'}      | ${SET_TODAY}      | ${date}             | ${setTodayPayload}
    ${'setVisibility'} | ${SET_VISIBILITY} | ${date}             | ${setVisibilityPayload}
    ${'setPrecision'}  | ${SET_PRECISION}  | ${precisionPayload} | ${precisionPayload}
  `('matchDatesWithPrecision', ({action, type, value, expectedPayload}) => {
    it(`${action} creator dispatches to reducer`, () => {
      const {result} = renderHook(() => useDatePickerActions(), {
        wrapper: ({children}) => <Provider>{children}</Provider>,
      });

      act(() => {
        result.current[action](value);
      });

      expect(mockReducer).toHaveBeenCalledTimes(1);

      expect(mockReducer.mock.calls[0][1]).toStrictEqual({
        payload: expectedPayload,
        type,
      });
    });
  });
});
