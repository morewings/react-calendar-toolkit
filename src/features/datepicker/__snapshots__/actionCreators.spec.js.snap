// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useModalActions returns action creators 1`] = `
Object {
  "setDate": [Function],
  "setPrecision": [Function],
  "setToday": [Function],
  "setVisibility": [Function],
}
`;
